File: problems/9 - Palindrome Number/tests/test_solution.ts
import * as fs from 'fs';
import * as path from 'path';

// Import the solution from the parent directory
const solution = require('../solution');

// Define interfaces and types for the test cases
interface TestCase {
  name: string;
  input: { x: number };
  expected: boolean;
}

// Helper method to convert between arrays and ListNode
function arrayToListNode(arr: number[]): ListNode | null {
  let head = new ListNode(arr[0]);
  let current = head;

  for (let i = 1; i < arr.length; i++) {
    const node = new ListNode(arr[i]);
    current.next = node;
    current = node;
  }

  return head;
}

function listNodeToArray(head: ListNode | null): number[] {
  const result: number[] = [];
  let current = head;

  while (current !== null) {
    result.push(current.val);
    current = current.next;
  }

  return result;
}

// Read JSON input from command line argument
const inputFile = process.argv[2];
if (!inputFile) {
  console.error('Please provide the input file path as a command line argument');
  process.exit(1);
}

const inputJson = fs.readFileSync(inputFile, 'utf8');
const inputData: TestCase[] = JSON.parse(inputJson);

// Call the solution method
for (const testCase of inputData) {
  const result = solution.isPalindrome(testCase.input.x);
  console.log(`Test case ${testCase.name}: Expected ${testCase.expected}, got ${result}`);
}

function ListNode(val?: number, next?: ListNode | null) {
  this.val = val === undefined ? 0 : val;
  this.next = next === undefined ? null : next;
}

File: problems/9 - Palindrome Number/tests/test_solution.py
import json
from pathlib import Path
import sys

# Helper methods to convert between arrays and ListNode
def array_to_list_node(arr):
    dummy = ListNode(0)
    current = dummy
    for num in arr:
        current.next = ListNode(num)
        current = current.next
    return dummy.next

def list_node_to_array(node):
    result = []
    while node:
        result.append(node.val)
        node = node.next
    return result

# Read JSON input from command line argument
if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage: python test_file.py <input_json>")
        sys.exit(1)

    with open(sys.argv[1], 'r') as f:
        input_data = json.load(f)

# Import the solution from the parent directory
sys.path.insert(0, str(Path(__file__).parent))
from palindrome_number import Solution

solution = Solution()

# Call the solution method and output JSON result
result = []
for test_case in input_data['testCases']:
    x = test_case['input']['x']
    expected = test_case['expected']
    actual = solution.isPalindrome(x)
    if actual == expected:
        result.append(1)
    else:
        result.append(0)

print(json.dumps({"result": result}))
